akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -883,24 +910,124 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
+
+    /**
+     * Triggers "graceful" termination request being handled on this connection.
+     *
+     * Termination works as follows:
+     *
+     * 1) Unbind:
+     * - the server port is unbound; no new connections will be accepted.
+     *
+     * 1.5) Immediately the ServerBinding `whenTerminationSignalIssued` future is completed.
+     * This can be used to signal parts of the application that the http server is shutting down and they should clean up as well.
+     * Note also that for more advanced shut down scenarios you may want to use the Coordinated Shutdown capabilities of Akka.
+     *
+     * 2) Handle in-flight request:
+     * - if a request is "in-flight" (being handled by user code), it is given `hardDeadline` time to complete,
+     *   - if user code emits a response within the timeout, then this response is sent to the client
+     *     - however if it is a streaming response, it is also mandated that it shall complete within the deadline, and if it does not
+     *       the connection will be terminated regardless of status of the streaming response (this is because such response could be infinite,
+     *       which could trap the server in a situation where it could not terminate if it were to wait for a response to "finish")
+     *   - if user code does not reply with a response within the deadline, we produce a special [[ServerSettings.terminationDeadlineExceededResponse]]
+     *     HTTP response (e.g. 503 Service Unavailable)
+     *
+     * 3) Keep draining incoming requests on existing connection:
+     * - The existing connection will remain alive for until the `hardDeadline` is exceeded,
+     *   yet no new requests will be delivered to the user handler. All such drained responses will be replied to with an
+     *   termination response (as explained in phase 2).
+     *
+     * 3) Close still existing connections
+     * - Connections are terminated forcefully once the `hardDeadline` is exceeded.
+     *   The `whenTerminated` future is completed as well, so the graceful termination (of the `ActorSystem` or entire JVM
+     *   itself can be safely performed, as by then it is known that no connections remain alive to this server).
+     *
+     * Note that the termination response is configurable in [[ServerSettings]], and by default is an `503 Service Unavailable`,
+     * with an empty response entity.
+     *
+     * @param hardDeadline timeout after which all requests and connections shall be forcefully terminated
+     * @return future which completes successfully with a marker object once all connections have been terminated
+     */
+    def terminate(hardDeadline: FiniteDuration): Future[HttpTerminated] = {
+      require(hardDeadline.toMillis > 0, "deadline must be greater than 0ms")
+
+      _whenTerminationSignalIssued.trySuccess(hardDeadline.fromNow)
+      val terminated = unbindAction().flatMap(_ ⇒ terminateAction(hardDeadline))(ExecutionContexts.sameThreadExecutionContext)
```

Comment:
sure it will never throw?

Reply:
I think unbind is more possible to throw perhaps... I'll require an execution context then